File: src/core/anchor_grid.cpp
#include "anchor_grid.h"

#include <algorithm>
#include <cstdint>
#include <map>

namespace {
int cell_id(float u, float v, int grid, float w, float h) {
    int cu = static_cast<int>(u * grid / w);
    int cv = static_cast<int>(v * grid / h);
    cu = cu < 0 ? 0 : (cu > grid - 1 ? grid - 1 : cu);
    cv = cv < 0 ? 0 : (cv > grid - 1 ? grid - 1 : cv);
    return cv * grid + cu;
}

void pick_cell(const std::vector<size_t>& members, const std::vector<float>& depth,
               int k, const std::string& pick, std::vector<size_t>& out) {
    const int n = static_cast<int>(members.size());
    if (n <= k) {
        out.insert(out.end(), members.begin(), members.end());
        return;
    }
    if (k == 1) {
        if (pick == "nearest") {
            out.push_back(*std::min_element(
                members.begin(), members.end(),
                [&](size_t a, size_t b) { return depth[a] < depth[b]; }));
        } else if (pick == "median") {
            std::vector<size_t> s = members;
            std::sort(s.begin(), s.end(),
                      [&](size_t a, size_t b) { return depth[a] < depth[b]; });
            out.push_back(s[n / 2]);
        } else {
            out.push_back(members[0]);  // first
        }
        return;
    }
    for (int i = 0; i < k; ++i) {
        out.push_back(members[static_cast<int>(
            static_cast<int64_t>(i) * (n - 1) / (k - 1))]);
    }
}
}  // namespace

void grid_thin(
    const std::vector<float>& u, const std::vector<float>& v,
    const std::vector<float>& depth, int grid, int k, float width, float height,
    const std::string& pick, std::vector<size_t>& keep) {
    std::map<int, std::vector<size_t>> by;  // ascending cell order
    for (size_t i = 0; i < u.size(); ++i) {
        by[cell_id(u[i], v[i], grid, width, height)].push_back(i);
    }
    for (const auto& cell : by) {
        pick_cell(cell.second, depth, k, pick, keep);
    }
}

File: proto/rescaler/anchor_grid.py
"""Grid-thin ToF anchors: keep up to K per GxG image cell."""
import numpy as np


def _cell_ids(uv: np.ndarray, grid: int, width: int, height: int) -> np.ndarray:
    if len(uv) == 0:
        return np.empty(0, dtype=np.int64)
    cu = np.clip((uv[:, 0] * grid / width).astype(np.int64), 0, grid - 1)
    cv = np.clip((uv[:, 1] * grid / height).astype(np.int64), 0, grid - 1)
    return cv * grid + cu


def _pick(members: list, depth: np.ndarray, k: int, rule: str) -> list:
    n = len(members)
    if n <= k:
        return members
    if k == 1:
        if rule == "nearest":
            return [min(members, key=lambda i: depth[i])]
        if rule == "median":
            s = sorted(members, key=lambda i: depth[i])
            return [s[n // 2]]
        return [members[0]]  # first (input order)
    return [members[int(i * (n - 1) / (k - 1))] for i in range(k)]


def grid_thin(uv: np.ndarray, depth: np.ndarray, grid: int, k: int,
              width: int, height: int, pick: str = "nearest"):
    """Indices into uv: up to k per cell. pick (k=1): first | nearest | median."""
    by: dict = {}
    for i, c in enumerate(_cell_ids(uv, grid, width, height)):
        by.setdefault(int(c), []).append(i)
    keep: list = []
    for c in sorted(by):
        keep += _pick(by[c], depth, k, pick)
    return np.array(keep, dtype=int)
